Workaround hcc bug regarding extern "C" definitions

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30315 Split libtorch.so back into libtorch_{cpu,cuda,hip}
* #30314 Add missing trigramma_stub definition.
* **#30313 Workaround hcc bug regarding extern "C" definitions**
* #30312 Delete redundant THC_API on THCStorage_new
* #30311 Properly include declaration of dispatch in file that registers it.
* #30310 Add missing _API definitions.
* #30308 DEFINE_DISPATCH in the correct namespace.
* #30307 Use normal dispatch to get to CUDA threshold kernels, instead of DispatchStub.

See comments in code about the bug.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18762360](https://our.internmc.facebook.com/intern/diff/D18762360)

ezyang committed:
Workaround hcc bug regarding extern "C" definitions

See comments in code about the bug.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bddppq commented on aten/src/THC/THCTensorRandom.cu:
```diff
@@ -37,7 +40,7 @@ THC_API __host__ void THCRandom_getRNGState(at::Generator *gen_, THByteTensor *r
 }
 
-THC_API __host__ void THCRandom_setRNGState(at::Generator *gen_, THByteTensor *rng_state)
+extern "C" __host__ void THCRandom_setRNGState(at::Generator *gen_, THByteTensor *rng_state)
```
https://github.com/pytorch/pytorch/blob/1d0d9b10ea7a4385947402ed0628ae1bace1c059/aten/src/THC/THCGeneral.h.in#L21
https://github.com/pytorch/pytorch/blob/1d0d9b10ea7a4385947402ed0628ae1bace1c059/c10/macros/Export.h#L105
How about c10 export part?

ezyang replied:
It's handled in the header, and that annotation is properly propagated.

ezyang committed:
Update on "Workaround hcc bug regarding extern "C" definitions"


See comments in code about the bug.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Workaround hcc bug regarding extern "C" definitions"


See comments in code about the bug.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18762360](https://our.internmc.facebook.com/intern/diff/D18762360)

[ghstack-poisoned]

ezyang committed:
Update on "Workaround hcc bug regarding extern "C" definitions"


See comments in code about the bug.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18762360](https://our.internmc.facebook.com/intern/diff/D18762360)

[ghstack-poisoned]

ezyang committed:
Update on "Workaround hcc bug regarding extern "C" definitions"


See comments in code about the bug.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18762360](https://our.internmc.facebook.com/intern/diff/D18762360)

[ghstack-poisoned]
